File: js/algorithms/paging/page_table.js
/*
 * page_table.js
 * Flat page table with TLB-assisted address translation.
 * Each virtual page maps to a frame. Page fault when valid=false.
 * JS port of c/algorithms/paging/page_table.c and tlb.c
 */

export function pageTableInit(pageSize, numPages) {
  const entries = [];
  for (let i = 0; i < numPages; i++) {
    entries.push({
      pageNumber: i,
      frameNumber: i,
      valid: true,
      dirty: false,
      referenced: false,
      accessCount: 0,
      lastAccessTime: -1
    });
  }
  return {
    entries,
    pageCount: numPages,
    pageSize,
    faultCount: 0,
    accessCount: 0
  };
}

export function loadPage(pt, pageNumber, frameNumber) {
  while (pt.entries.length <= pageNumber) {
    pt.entries.push({
      pageNumber: pt.entries.length,
      frameNumber: -1,
      valid: false,
      dirty: false,
      referenced: false,
      accessCount: 0,
      lastAccessTime: -1
    });
  }
  pt.entries[pageNumber].frameNumber = frameNumber;
  pt.entries[pageNumber].valid = true;
  if (pageNumber >= pt.pageCount) pt.pageCount = pageNumber + 1;
}

export function invalidatePage(pt, pageNumber) {
  if (pageNumber < pt.entries.length) {
    pt.entries[pageNumber].valid = false;
  }
}

export function translateAddress(pt, tlb, logicalAddress) {
  const pageNumber = Math.floor(logicalAddress / pt.pageSize);
  const offset = logicalAddress % pt.pageSize;
  const tr = {
    logicalAddress,
    pageNumber,
    offset,
    frameNumber: -1,
    physicalAddress: -1,
    pageFault: false,
    tlbHit: false
  };

  const tlbResult = tlbLookup(tlb, pageNumber);
  if (tlbResult !== -1) {
    tr.tlbHit = true;
    tr.frameNumber = tlbResult;
  } else {
    if (pageNumber < pt.entries.length && pt.entries[pageNumber].valid) {
      tr.frameNumber = pt.entries[pageNumber].frameNumber;
      tlbInsert(tlb, pageNumber, tr.frameNumber);
    } else {
      tr.pageFault = true;
      pt.faultCount++;
      pt.accessCount++;
      return tr;
    }
  }

  tr.physicalAddress = tr.frameNumber * pt.pageSize + offset;
  pt.entries[pageNumber].accessCount++;
  pt.entries[pageNumber].referenced = true;
  pt.entries[pageNumber].lastAccessTime = pt.accessCount;
  pt.accessCount++;
  return tr;
}

/* ===== TLB ===== */
export function tlbInit() {
  return {
    entries: new Array(16).fill(null).map(() => ({
      pageNumber: -1, frameNumber: -1, valid: false, lastUsed: -1
    })),
    hits: 0,
    misses: 0,
    time: 0
  };
}

export function tlbLookup(tlb, pageNumber) {
  tlb.time++;
  for (const e of tlb.entries) {
    if (e.valid && e.pageNumber === pageNumber) {
      e.lastUsed = tlb.time;
      tlb.hits++;
      return e.frameNumber;
    }
  }
  tlb.misses++;
  return -1;
}

export function tlbInsert(tlb, pageNumber, frameNumber) {
  tlb.time++;
  let victim = tlb.entries.findIndex(e => !e.valid);
  if (victim === -1) {
    victim = 0;
    let minT = tlb.entries[0].lastUsed;
    for (let i = 1; i < tlb.entries.length; i++) {
      if (tlb.entries[i].lastUsed < minT) { minT = tlb.entries[i].lastUsed; victim = i; }
    }
  }
  tlb.entries[victim] = { pageNumber, frameNumber, valid: true, lastUsed: tlb.time };
}

export function tlbFlush(tlb) {
  for (const e of tlb.entries) { e.valid = false; e.pageNumber = -1; }
  tlb.hits = 0;
  tlb.misses = 0;
}

export function tlbHitRatio(tlb) {
  const total = tlb.hits + tlb.misses;
  return total === 0 ? 0 : tlb.hits / total;
}

File: c/algorithms/paging/page_table.c
/*
 * page_table.c
 * Flat page table implementation with TLB-assisted address translation.
 * Each virtual page maps to a frame. A page fault occurs when the valid
 * bit is 0. TLB is checked first for fast translation.
 * Time complexity: O(1) TLB hit, O(n) TLB miss + lookup.
 */

#include "page_table.h"
#include "tlb.h"
#include <string.h>

void page_table_init(PageTable *pt, int page_size, int num_pages) {
    memset(pt, 0, sizeof(PageTable));
    pt->page_size = page_size;
    pt->page_count = num_pages;
    int i;
    for (i = 0; i < num_pages && i < MAX_PAGES; i++) {
        pt->entries[i].page_number = i;
        pt->entries[i].frame_number = i; /* default 1:1 mapping */
        pt->entries[i].valid = 1;
        pt->entries[i].dirty = 0;
        pt->entries[i].referenced = 0;
        pt->entries[i].access_count = 0;
        pt->entries[i].last_access_time = -1;
    }
}

AddressTranslation translate_address(PageTable *pt, TLB *tlb, int logical_address) {
    AddressTranslation tr;
    memset(&tr, 0, sizeof(AddressTranslation));
    tr.logical_address = logical_address;
    tr.page_number = logical_address / pt->page_size;
    tr.offset = logical_address % pt->page_size;

    int frame = -1;
    if (tlb_lookup(tlb, tr.page_number, &frame)) {
        tr.tlb_hit = 1;
        tr.frame_number = frame;
    } else {
        tr.tlb_hit = 0;
        if (tr.page_number < pt->page_count && pt->entries[tr.page_number].valid) {
            tr.frame_number = pt->entries[tr.page_number].frame_number;
            tlb_insert(tlb, tr.page_number, tr.frame_number);
        } else {
            tr.page_fault = 1;
            pt->fault_count++;
            tr.frame_number = -1;
            tr.physical_address = -1;
            pt->access_count++;
            return tr;
        }
    }

    tr.physical_address = tr.frame_number * pt->page_size + tr.offset;
    pt->entries[tr.page_number].access_count++;
    pt->entries[tr.page_number].referenced = 1;
    pt->access_count++;
    return tr;
}

void load_page(PageTable *pt, int page_number, int frame_number) {
    if (page_number >= MAX_PAGES) return;
    pt->entries[page_number].page_number = page_number;
    pt->entries[page_number].frame_number = frame_number;
    pt->entries[page_number].valid = 1;
    if (page_number >= pt->page_count) pt->page_count = page_number + 1;
}

void invalidate_page(PageTable *pt, int page_number) {
    if (page_number >= MAX_PAGES) return;
    pt->entries[page_number].valid = 0;
}
